File: src/ssh_manager/util/paths.py
"""Home-dir resolution + .env loading.

Resolution (highest precedence first):
  1. ``$SSH_MANAGER_HOME`` (alias ``$SSH_MANAGER_CONFIG_DIR``), absolutized - explicit override.
  2. the OS-standard config dir + single app folder ``ssh-manager`` (``platform.config_dir()``):
       Unix/macOS: ``$XDG_CONFIG_HOME/ssh-manager`` if set, else ``~/.config/ssh-manager``
       Windows:    ``%APPDATA%\\ssh-manager``
A legacy ``~/.sshmgr`` home is auto-migrated to (2) on first run (see
``SshManagerService._migrate_legacy_home``). There is no project-local mode - one home
is used everywhere. The :class:`Paths` bundle is the single place layout is
computed, so every service agrees on where state lives.

One folder holds all of a user's state - config, secrets, logs, snapshots:

    <home>/
    |-- manifest.json inventory.json     (manifest is the source of truth)
    |-- providers.json                   (optional; else the shipped default catalog)
    |-- .env  age-identity.txt           (secrets, 0600)
    |-- log/audit.log
    |-- snapshots/
    |-- dist/    (exported encrypted bundles - ssh-manager-YYYYMMDD.age)
    `-- .state/  (.lock, expiry-cache.json, notify-cache.json)
"""
from __future__ import annotations

import os
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path

from ..platforms.base import Platform


@dataclass(frozen=True)
class Paths:
    """Resolved on-disk locations for one invocation."""

    ssh_dir: Path
    config_dir: Path        # the per-user home (OS-standard ssh-manager dir, or $SSH_MANAGER_HOME)

    @property
    def home(self) -> Path:
        return self.config_dir

    # config (the source of truth)
    @property
    def manifest(self) -> Path:
        return self.config_dir / "manifest.json"

    @property
    def inventory(self) -> Path:
        return self.config_dir / "inventory.json"

    @property
    def providers(self) -> Path:
        return self.config_dir / "providers.json"

    # secrets
    @property
    def env_file(self) -> Path:
        return self.config_dir / ".env"

    @property
    def env_example(self) -> Path:
        return self.config_dir / ".env-example"

    @property
    def age_identity(self) -> Path:
        return self.config_dir / "age-identity.txt"

    # logs
    @property
    def log_dir(self) -> Path:
        return self.config_dir / "log"

    @property
    def audit_log(self) -> Path:
        return self.log_dir / "audit.log"

    # backups
    @property
    def snapshots_dir(self) -> Path:
        return self.config_dir / "snapshots"

    # exported artifacts (encrypted bundles) live under the home, never the cwd
    @property
    def dist_dir(self) -> Path:
        return self.config_dir / "dist"

    # transient state
    @property
    def state_dir(self) -> Path:
        return self.config_dir / ".state"

    @property
    def lock_file(self) -> Path:
        return self.state_dir / ".lock"

    @property
    def expiry_cache(self) -> Path:
        return self.state_dir / "expiry-cache.json"

    @property
    def notify_cache(self) -> Path:
        return self.state_dir / "notify-cache.json"


def resolve_config_dir(platform: Platform, *,
                       env: dict[str, str] | None = None,
                       cwd: Path | None = None) -> Path:
    """Resolve the per-user home: ``$SSH_MANAGER_HOME`` / ``$SSH_MANAGER_CONFIG_DIR``
    (absolutized) if set, else the OS-standard config dir + ``ssh-manager`` folder
    (``platform.config_dir()`` -> ``~/.config/ssh-manager`` on Unix/macOS,
    ``%APPDATA%\\ssh-manager`` on Windows). No project-local ``./config`` mode; a legacy
    ``~/.sshmgr`` is auto-migrated by the service on first run."""
    source: Mapping[str, str] = os.environ if env is None else env
    override = source.get("SSH_MANAGER_HOME") or source.get("SSH_MANAGER_CONFIG_DIR")
    if override:
        p = Path(override).expanduser()
        if not p.is_absolute():
            p = (cwd or Path.cwd()) / p
        return p
    return platform.config_dir(env=source)


def resolve_paths(platform: Platform, *,
                  env: dict[str, str] | None = None,
                  cwd: Path | None = None,
                  ssh_dir: Path | None = None) -> Paths:
    """Build the full :class:`Paths` bundle for this invocation."""
    config_dir = resolve_config_dir(platform, env=env, cwd=cwd)
    return Paths(ssh_dir=ssh_dir or platform.ssh_dir(), config_dir=config_dir)


def load_env(env_file: Path) -> None:
    """Load the ``.env`` into the process env if present (best-effort)."""
    if not env_file.exists():
        return
    try:
        from dotenv import load_dotenv
    except ImportError:  # pragma: no cover - dotenv is a hard dep, defensive
        return
    load_dotenv(env_file, override=False)

File: internal/util/paths/paths.go
// Package paths resolves the per-user home and the on-disk layout, ported from
// src/ssh_manager/util/paths.py + the platform config_dir logic. Resolution:
// $SSH_MANAGER_HOME (alias $SSH_MANAGER_CONFIG_DIR), else the OS-standard config
// dir + the "ssh-manager" folder ($XDG_CONFIG_HOME or ~/.config on Unix/macOS,
// %APPDATA% on Windows).
package paths

import (
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

// Getenv is an env lookup, injectable for tests. nil means os.Getenv.
type Getenv func(string) string

func resolve(get Getenv) Getenv {
	if get == nil {
		return os.Getenv
	}
	return get
}

func home(get Getenv) string {
	if runtime.GOOS == "windows" {
		if v := get("USERPROFILE"); v != "" {
			return v
		}
	} else if v := get("HOME"); v != "" {
		return v
	}
	h, _ := os.UserHomeDir()
	return h
}

func expandUser(p string, get Getenv) string {
	if p == "~" {
		return home(get)
	}
	if strings.HasPrefix(p, "~/") || strings.HasPrefix(p, `~\`) {
		return filepath.Join(home(get), p[2:])
	}
	return p
}

// ConfigDir resolves the per-user home directory. cwd is used to absolutize a
// relative $SSH_MANAGER_HOME override (pass "" to use the process cwd).
func ConfigDir(get Getenv, cwd string) string {
	get = resolve(get)
	if override := firstNonEmpty(get("SSH_MANAGER_HOME"), get("SSH_MANAGER_CONFIG_DIR")); override != "" {
		p := expandUser(override, get)
		if !filepath.IsAbs(p) {
			if cwd == "" {
				cwd, _ = os.Getwd()
			}
			p = filepath.Join(cwd, p)
		}
		return p
	}
	if runtime.GOOS == "windows" {
		base := get("APPDATA")
		if base == "" {
			base = filepath.Join(home(get), "AppData", "Roaming")
		}
		return filepath.Join(base, "ssh-manager")
	}
	base := get("XDG_CONFIG_HOME")
	if base == "" {
		base = filepath.Join(home(get), ".config")
	}
	return filepath.Join(base, "ssh-manager")
}

func firstNonEmpty(vals ...string) string {
	for _, v := range vals {
		if v != "" {
			return v
		}
	}
	return ""
}

// Paths is the resolved on-disk layout for one invocation.
type Paths struct {
	SSHDir    string
	ConfigDir string
}

// Resolve builds the Paths bundle. sshDir "" defaults to ~/.ssh.
func Resolve(get Getenv, cwd, sshDir string) Paths {
	get = resolve(get)
	if sshDir == "" {
		sshDir = filepath.Join(home(get), ".ssh")
	}
	return Paths{SSHDir: sshDir, ConfigDir: ConfigDir(get, cwd)}
}

func (p Paths) Home() string         { return p.ConfigDir }
func (p Paths) Manifest() string     { return filepath.Join(p.ConfigDir, "manifest.json") }
func (p Paths) Inventory() string    { return filepath.Join(p.ConfigDir, "inventory.json") }
func (p Paths) Providers() string    { return filepath.Join(p.ConfigDir, "providers.json") }
func (p Paths) EnvFile() string      { return filepath.Join(p.ConfigDir, ".env") }
func (p Paths) AgeIdentity() string  { return filepath.Join(p.ConfigDir, "age-identity.txt") }
func (p Paths) LogDir() string       { return filepath.Join(p.ConfigDir, "log") }
func (p Paths) AuditLog() string     { return filepath.Join(p.LogDir(), "audit.log") }
func (p Paths) SnapshotsDir() string { return filepath.Join(p.ConfigDir, "snapshots") }
func (p Paths) DistDir() string      { return filepath.Join(p.ConfigDir, "dist") }
func (p Paths) StateDir() string     { return filepath.Join(p.ConfigDir, ".state") }
func (p Paths) LockFile() string     { return filepath.Join(p.StateDir(), ".lock") }
func (p Paths) ExpiryCache() string  { return filepath.Join(p.StateDir(), "expiry-cache.json") }
func (p Paths) NotifyCache() string  { return filepath.Join(p.StateDir(), "notify-cache.json") }
